platforms/software/dependency-management/src/main/java/org/gradle/internal/component/resolution/failure/ReportableAsProblem.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.component.resolution.failure;
+
+import org.gradle.api.problems.internal.InternalProblems;
+
+/**
+ * Represents a {@link Throwable} that can be reported as a {@link org.gradle.api.problems.internal.Problem} to
+ * the {@link org.gradle.api.problems.Problems} service.
+ * <p>
+ * This is necessary, as Dependency Management uses exceptions for flow control, and some exceptions are meant to
```

Comment:
🤔 What do you mean here by `Dependency Management uses exceptions for flow control`. I'd argue the opposite, where we _don't_ use exceptions for flow control but rather eagerly capture exceptions and hold them as state, only throwing them under certain conditions.

Reply:
I do mean exactly we throw exceptions under certain (rather routine, and easy to anticipate) conditions that contain the failure information, and are later caught without failing the build rather than just returning some of failure type.

I'm just anticipating a reader wondering why we don't just report the problem immediately.  Open to suggestions to improve the wording, but on re-reading I still think this makes sense.